versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractTieredCommitsTest.java

```diff
@@ -0,0 +1,1193 @@
+    databaseAdapter.toHash(main);
+  }
+
+  static class ManyKeysParams {
+    final int keys;
+    final int commits;
+
+    public ManyKeysParams(int keys, int commits) {
+      this.keys = keys;
+      this.commits = commits;
+    }
+
+    @Override
+    public String toString() {
+      return "keys=" + keys + ", commits=" + commits;
+    }
+  }
+
+  static List<ManyKeysParams> manyKeysParams() {
+    return Arrays.asList(
+        // quite slow for a unit-test
+        // new ManyKeysParams(20000, 25),
+        // new ManyKeysParams(20000, 100),
+        new ManyKeysParams(250, 25),
+        new ManyKeysParams(1000, 25),
+        new ManyKeysParams(1000, 100),
+        new ManyKeysParams(5000, 25),
+        new ManyKeysParams(5000, 100));
+  }
+
+  @ParameterizedTest
+  @MethodSource("manyKeysParams")
+  void manyKeys(ManyKeysParams params) throws Exception {
+    BranchName main = BranchName.of("main");
+
+    List<ImmutableCommitAttempt.Builder> commits =
+        IntStream.range(0, params.commits)
+            .mapToObj(
+                i ->
+                    ImmutableCommitAttempt.builder()
+                        .commitMetaSerialized(ByteString.copyFromUtf8("commit #" + i))
+                        .commitToBranch(main))
+            .collect(Collectors.toList());
+    AtomicInteger commitDist = new AtomicInteger();
+
+    Set<Key> allKeys = new HashSet<>();
+
+    IntStream.range(0, params.keys)
+        .mapToObj(
+            i -> {
+              Key key =
+                  Key.of(
+                      "some",
+                      Integer.toString(i),
+                      "long",
+                      "key",
+                      "value",
+                      "foobarbazfoobarbazfoobarbazfoobarbazfoobarbazfoobarbaz");
+              allKeys.add(key);
+              return KeyWithBytes.of(
+                  key, ContentsId.of("cid-" + i), (byte) 0, ByteString.copyFromUtf8("value " + i));
+            })
+        .forEach(kb -> commits.get(commitDist.incrementAndGet() % params.commits).addPuts(kb));
+
+    for (ImmutableCommitAttempt.Builder commit : commits) {
+      databaseAdapter.commit(commit.build());
+    }
+
+    Hash mainHead = databaseAdapter.toHash(main);
+    try (Stream<KeyWithType> keys = databaseAdapter.keys(mainHead, KeyFilterPredicate.ALLOW_ALL)) {
+      List<Key> fetchedKeys = keys.map(KeyWithType::getKey).collect(Collectors.toList());
+
+      // containsExactlyInAnyOrderElementsOf() is quite expensive and slow with Key's
+      // implementation of 'Key.equals()' since it uses a collator.
+      List<String> fetchedKeysStrings =
+          fetchedKeys.stream().map(Key::toString).collect(Collectors.toList());
+      List<String> allKeysStrings =
+          allKeys.stream().map(Key::toString).collect(Collectors.toList());
+
+      assertThat(fetchedKeysStrings)
+          .hasSize(allKeysStrings.size())
+          .containsExactlyInAnyOrderElementsOf(allKeysStrings);
+    }
+  }
+}
```

Comment:
1200 lines for a single test class is quite a lot :)

Reply:
smaller now